File: go/problems/83_remove_duplicates_from_sorted_list_test.go
package main

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestDeleteDuplicates(t *testing.T) {
	tests := []struct {
		head []int
		want []int
	}{
		{
			head: []int{1, 1, 2},
			want: []int{1, 2},
		},
		{
			head: []int{1, 1, 2, 3, 3},
			want: []int{1, 2, 3},
		},
		{
			head: []int{1, 1, 1},
			want: []int{1},
		},
	}

	for _, tt := range tests {
		t.Run(fmt.Sprintf("provided %v, we expect %v", tt.head, tt.want), func(t *testing.T) {
			got := toInt(deleteDuplicates(toList(tt.head)))
			assert.Equal(t, tt.want, got, "got %v want %v", got, tt.want)
		})
	}
}

File: ts/problems/83-remove-duplicates-from-sorted-list.test.ts
import { describe, expect, test } from "vitest";
import {
  deleteDuplicates,
  type ListNode,
} from "./83-remove-duplicates-from-sorted-list";

function toList(nums: Array<number>): ListNode | null {
  if (!nums.length) return null;
  return nums.reduceRight<ListNode | null>(
    (next, val) => ({ val, next }),
    null,
  );
}

describe("deleteDuplicates", () => {
  test.each([
    {
      head: [1, 1, 2],
      want: [1, 2],
    },
    {
      head: [1, 1, 2, 3, 3],
      want: [1, 2, 3],
    },
    {
      head: [1, 1, 1],
      want: [1],
    },
  ])("provided $head, we expect $want", ({ head, want }) => {
    expect(deleteDuplicates(toList(head))).toStrictEqual(toList(want));
  });
});
